Refactor control handling

sellist opened the pull request:
Hey!

During my overarching idea of creating a `honk` action into multiplayer, I found myself getting stuck on the control handling. With no centralized place to handle all controls, there were some difficulties to adding more controls into UI: some keys were used for hardcoded functions in the UI, line-number-based file persistence, and handling keycodes.

This MR creates the CurrentControls class which provides an instance of the CurrentControls (held in CurrentControlsInstanceHolder) to any requiring class, and all key information is provided by a Hotkey class which holds any requiring details (keycode, label, config prefix)

File persistence is performed by the stdlib Properties class, to save manually adding line numbers.

I tried going through the code to address any hard coding of keys, but I may have missed some.

sellist committed:
refactor settings

sellist committed:
Merge branch 'jhoffis:main' into refactor-settings

sellist committed:
consolidate line numbers to static variables

sellist committed:
Merge remote-tracking branch 'origin/refactor-settings' into refactor-settings

sellist committed:
update scenes with new control refactor

sellist committed:
finish up

jhoffis commented:
I do prefer the isKey(key) solution that already was in place in Race, as it is easier to debug perhaps, but it's fine 👍. The only thing is that the keys for gears like NUJIKOL is really old lol. It's from Update 3 and earlier. So that can be removed I think. Another thing is the Hotkey repeats strings twice. I see that the only real difference is spacing. I would recommend using enums and then using the function (which i don't completely remember where is) which will recognize upper letters of enums and return that, I use it for all the names of tiles, so it might be somewhere with TileNames, or Texts, or TileVisual perhaps. This would allow it to recognize on compiletime if you wrote a name wrong and you would only need to write it once in the hotkey constructor.

jhoffis commented:
I also added a new branch which maintains the current update's code untouched, so you can more freely change the main branch with pr's :)

The enum function I mentioned above is called "public static String getUpgradeTitle(TileNames imageType)" and in Texts.java at line 210.

jhoffis commented:
It would probably be good to generalize this function like for instance "createSpacedEnumString(Enum enum)" instead?

jhoffis commented:
Another issue I now see is that controls.refresh is being called every frame in RaceVisual. This calls upon the disk many times. This is really really slow. Either this should be removed entirely, or moved to initWindow in Race or to when the program starts up and when the hotkeys are changed in the options menu, right?

jhoffis commented:
Another thing again, lol, is that I see that we have doubled up 	private final CurrentControls controls = CurrentControls.getInstance(); in both UpgradesSubscene and Race. This is probably something I would write in the past too, and this pattern is all over the place. But as I've learnt over time the better solution is to just to use a singleton in this specific scenario. 
For example, RSet used to be much worse in the past, and was like this where it had to referenced all over the place but it was hard to reference, but now it's static and so much easier to use.
The way CurrentControls is used in Lobby is perhaps better? Just thinking out loud. Because then there's only one point of failure rather than 3+ points.

Here's an idea of something else: Perhaps the hotkeys in CurrentControls should be like RSet where you have a list of Hotkeys and then an enum which decides the names and types all in one. This would require much less writing of setters and getters and intialization and is much easier to expand in the future.


`
public class CurrentControls {

    static enum HotkeyNames {
        throttle, brake etc.
    };

    private final Hotkey[] keys;
    
    protected CurrentControls() {
        keys[HotkeyNames.throttle.toInt] = new Hotkey(HotkeyNames.throttle, GLFW.GLFW_KEY_W);
        keys[HotkeyNames.brake.toInt] = new Hotkey(HotkeyNames.brake, GLFW.GLFW_KEY_S);
    }
`

sellist commented:
Do agree on all your points given, will work on addressing your comments!

Also, I'm not super familiar with Github PRs, were your comments supposed to be on certain line numbers? In my view it's just in the PR itself, which I can understand all the meanings and references, but am unsure if I'm missing a view. Thank you!

jhoffis commented:
No, i was just commenting generally👍

sellist committed:
switch to enum instance for singleton rather than static provider

sellist committed:
clean up controls and hotkey classes

sellist committed:
change "== getKeycode" pattern to be more streamlined, updated usages

sellist committed:
remove auto generated hashcode method

sellist committed:
switch to more property access usage rather than getters, remove old debug sout

sellist committed:
remove unused argument

sellist committed:
initialize keycode on init

sellist commented:
Did the change to the enum singleton rather than getting an instance, much better pattern, will have to remember that! Also had the benefit for fixing up some other issues you mentioned.

Went about having the properties file key auto generated based on the hotkey's name to eliminate that doubling up of the constructor args as well.

sellist committed:
add beep to car in singleplayer

sellist committed:
add beep to multiplayer

sellist committed:
Merge pull request #1 from sellist/add-beep

add-beep
